pkg/server/commands/batch_check_command.go

```diff
@@ -0,0 +1,141 @@
+	StoreID              string
+}
+
+type BatchCheckOutcome struct {
+	CheckResponse *graph.ResolveCheckResponse
+	Duration      time.Duration
+	Err           error
+}
+
+type BatchCheckQueryOption func(*BatchCheckQuery)
+
+func WithBatchCheckCommandCacheController(cc cachecontroller.CacheController) BatchCheckQueryOption {
+	return func(bq *BatchCheckQuery) {
+		bq.cacheController = cc
+	}
+}
+
+func WithBatchCheckCommandLogger(l logger.Logger) BatchCheckQueryOption {
+	return func(bq *BatchCheckQuery) {
+		bq.logger = l
+	}
+}
+
+func WithBatchCheckMaxConcurrentChecks(maxConcurrentChecks uint32) BatchCheckQueryOption {
+	return func(bq *BatchCheckQuery) {
+		bq.maxConcurrentChecks = maxConcurrentChecks
+	}
+}
+
+func WithBatchCheckMaxChecksPerBatch(maxChecks uint32) BatchCheckQueryOption {
+	return func(bq *BatchCheckQuery) {
+		bq.maxChecksAllowed = maxChecks
+	}
+}
+
+func NewBatchCheckCommand(datastore storage.RelationshipTupleReader, checkResolver graph.CheckResolver, typesys *typesystem.TypeSystem, opts ...BatchCheckQueryOption) *BatchCheckQuery {
+	cmd := &BatchCheckQuery{
+		logger:              logger.NewNoopLogger(),
+		datastore:           datastore,
+		checkResolver:       checkResolver,
+		typesys:             typesys,
+		maxChecksAllowed:    config.DefaultMaxChecksPerBatchCheck,
+		maxConcurrentChecks: config.DefaultMaxConcurrentChecksPerBatchCheck,
+	}
+
+	for _, opt := range opts {
+		opt(cmd)
+	}
+	return cmd
+}
+
+func (bq *BatchCheckQuery) Execute(ctx context.Context, params *BatchCheckCommandParams) (map[string]*BatchCheckOutcome, error) {
+	if len(params.Checks) > int(bq.maxChecksAllowed) {
+		return nil, errors.ValidationError(
+			fmt.Errorf("batchCheck received %d checks, the maximum allowed is %d ", len(params.Checks), bq.maxChecksAllowed),
+		)
+	}
+
+	// This check query will be run against every check in the batch
+	checkQuery := NewCheckCommand(
+		bq.datastore,
+		bq.checkResolver,
+		bq.typesys,
+		WithCheckCommandLogger(bq.logger),
+		WithCacheController(bq.cacheController),
+	)
+
+	// the keys to this map are the correlation_id associated with each check
+	var resultMap = map[string]*BatchCheckOutcome{}
+	lock := sync.Mutex{}
+
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	pool := concurrency.NewPool(ctx, int(bq.maxConcurrentChecks))
+	for _, check := range params.Checks {
+		pool.Go(func(ctx context.Context) error {
+			checkParams := &CheckCommandParams{
+				StoreID:          params.StoreID,
+				TupleKey:         check.GetTupleKey(),
+				ContextualTuples: check.GetContextualTuples(),
+				Context:          check.GetContext(),
+				Consistency:      params.Consistency,
+			}
+
+			start := time.Now()
+
+			response, _, err := checkQuery.Execute(ctx, checkParams)
+
+			lock.Lock()
+			resultMap[check.GetCorrelationId()] = &BatchCheckOutcome{
+				CheckResponse: response,
+				Duration:      time.Since(start),
+				Err:           err,
+			}
+			lock.Unlock()
+
+			return nil
```

Comment:
Could there be a case where we do want to return non-nil? How would something like `context.DeadlineExceeded` or `context.Canceled` behave here?

Reply:
Will and I actually talked about this the other day. How would we want something like that to behave here? Or would we want something like this at the top of each pool.Go:
```
select {
case <- ctx.Done():
  // return a timeout error from this routine
default:
  // do nothing and let the routine run its check
}
```

Reply:
My expectation is that the underlying check resolver handles `context.DeadlineExceeded` and `context.Canceled` errors. Each affected check would surface its own error. This allows partial results to be fulfilled. Besides validations, the batch check command is just an orchestrator that doesn't conduct any other operations that could result in an error.

Reply:
That makes sense. In that case, it should work as described above with no additional changes.